### ISSUE TITLE
Run tests in parallel

### ISSUE THREAD

**Issue by lewisob** [serenity-bdd/serenity-cucumber#148]
As a tester who needs to run lots of tests, it would be of great benefit if the cucumber serenity runner could run tests in parallel (using gradle).

I know there are some other solutions/hacks out there. But none of them are easy to setup and can result in multiple/broken serenity reports.

**Comment by wakaleo** [comment]
I'm not sure this is doable without rewriting large bits of the Cucumber code base, but I'm open to suggestions or pull requests.

**Comment by nbarrett** [comment]
The latest cucumber-serenity release now supports [parallel running of cucumber tests](https://serenity-bdd.github.io/theserenitybook/latest/serenity-parallel.html)

**Comment by lewisob** [comment]
Hi @nbarrett 
Is the parallel feature only for CI?

**Comment by nbarrett** [comment]
@lewisob - not sure what you mean 'only CI', but there is only documented information so far on how to execute the parallelism via maven at the moment.

**Comment by lewisob** [comment]
Hi @nbarrett the documentation mentions "when it was run in your CI environment" and "build slaves" a number of times so I thought the instructions may be specific to CI, e.g. a Jenkins build or something.

**Comment by nbarrett** [comment]
Hi again @lewisob - there's no reason why you can't run all of the batches on different machines without using ci to orchestrate it but it would be a bit more complex to copy all the artifacts manually afterwards - but perfectly possible. All the commands are regular maven commands that can be executed on any machine.

**Comment by lewisob** [comment]
Thanks @nbarrett 
Can I run in parallel on 1 machine? Seems like it's still a difficult problem to solve.

**Comment by nbarrett** [comment]
Yes you can if you use the maven configuration as described in [the serenity book](https://serenity-bdd.github.io/theserenitybook/latest/serenity-parallel.html#_subdividing_a_cucumber_batches_within_a_machine). However you need to make sure that your machine has the resources to be able to run the number of browsers specified in the `<serenity.fork.count>0${parallel.tests}</serenity.fork.count>` property. As a general rule, I wouldn't even recommend doing any parallel testing until I have a test suite that runs reliably (e.g. 100% green) on a single host. Otherwise it can be hard to diagnose problems that can crop up (especially as you will be sharing a single test environment with multiple clients that could easily be accessing/changing shared data that could cause test failures.

**Comment by lewisob** [comment]
Ok. I am using Gradle, so will need to convert it. I will attempt.
Thanks for your help.

**Comment by scormaq** [comment]
@lewisob I also use Gradle in my Cucumber-Serenity project. There is no problems with batching, but a bit of problem with forking (I did not found a way to reuse `maven-failsafe-plugin` mechanism in Gradle build script). Instead of it I managed to deal with forking in Gradle test parallelization mechanism. You can check my Serenity-Cucumber forking solution for Gradle here: https://github.com/scormaq/gradle-serenity-cucumber-forking

**Comment by nbarrett** [comment]
> @lewisob I also use Gradle in my Cucumber-Serenity project. There is no problems with batching, but a bit of problem with forking (I did not found a way to reuse `maven-failsafe-plugin` mechanism in Gradle build script). Instead of it I managed to deal with forking in Gradle test parallelization mechanism. You can check my Serenity-Cucumber forking solution for Gradle here: https://github.com/scormaq/gradle-serenity-cucumber-forking

This looks awesome @lewisob! Are you considering creating a PR to add this capability to serenity?

**Comment by scormaq** [comment]
@nbarrett The only thing that worth to be contributed is `SerenityCucumberForker`, the rest of code is anyway an extra configuration. But I'm not sure that `SerenityCucumberForker` is implemented in optimal way. I'm doing risky stuff there:
1. I am not 100% sure that gradle worker props always be serial;
2. I'm injecting serenity properties in runtime, but I think this is not the best practice 😄 

If you believe that this approach (attaching sliced formatter as cucumber plugin) is acceptable to be merged into main repo, I can propose a PR when I have time.

**Comment by mlvandijk** [comment]
Note that Cucumber supports parallel execution as of Cucumber-jvm v4.0.0 - see [CHANGELOG.md](https://github.com/cucumber/cucumber-jvm/blob/master/CHANGELOG.md), so this issue is a duplicate of / related to #178

**Comment by chronicideas** [comment]
@scormaq thanks for sharing your gradle solution much appreciated

**Comment by thomas0306** [comment]
@scormaq The example code is very helpful since most of the parallel execution examples were done with maven. Thank you!
